sdk/servicebus/azure-servicebus/azure/servicebus/_common/message.py

```diff
@@ -429,64 +306,124 @@ def __init__(self, message):
+    .. admonition:: Example:
 
+        .. literalinclude:: ../samples/sync_samples/sample_code_servicebus.py
+            :start-after: [START receive_complex_message]
+            :end-before: [END receive_complex_message]
+            :language: python
+            :dedent: 4
+            :caption: Checking the properties on a received message.
     """
-
-    def __init__(self, message, mode):
-        self._settled = mode == 0
-        super(DeferredMessage, self).__init__(None, message=message)
+    def __init__(self, message, mode=ReceiveSettleMode.PeekLock):
+        self._settled = (mode == ReceiveSettleMode.ReceiveAndDelete)
+        super(ReceivedMessage, self).__init__(message=message)
 
     def _is_live(self, action):
-        if not self._receiver:
-            raise ValueError("Orphan message had no open connection.")
-        super(DeferredMessage, self)._is_live(action)
+        # pylint: disable=no-member
+        if not self._receiver or not self._receiver._running:  # pylint: disable=protected-access
+            raise MessageSettleFailed(action, "Orphan message had no open connection.")
+        if self.settled:
+            raise MessageAlreadySettled(action)
+        try:
+            if self.expired:
+                raise MessageLockExpired(inner_exception=self.auto_renew_error)
+        except TypeError:
+            pass
+        try:
+            if self._receiver.session and self._receiver.session.expired:
+                raise SessionLockExpired(inner_exception=self._receiver.session.auto_renew_error)
+        except TypeError: #TODO: Exception: AttributeError?
+            pass
+
+    @property
+    def settled(self):
+        """Whether the message has been settled.
+
+        This will aways be `True` for a message received using ReceiveAndDelete mode,
+        otherwise it will be `False` until the message is completed or otherwise settled.
+
+        :rtype: bool
+        """
+        return self._settled
+
+    @property
+    def expired(self):
+        if self._receiver._session_id:  # pylint: disable=protected-access
+            raise TypeError("Session messages do not expire. Please use the Session expiry instead.")
+        if self.locked_until_utc and self.locked_until_utc <= utc_now():
+            return True
+        return False
+
+    @property
+    def locked_until_utc(self):
+        if self._receiver._session_id or self.settled:  # pylint: disable=protected-access
+            return None
+        if self._expiry:
+            return self._expiry
+        if self.message.annotations and _X_OPT_LOCKED_UNTIL in self.message.annotations:
+            expiry_in_seconds = self.message.annotations[_X_OPT_LOCKED_UNTIL]/1000
+            self._expiry = utc_from_timestamp(expiry_in_seconds)
+        return self._expiry
 
     @property
     def lock_token(self):
         if self.settled:
             return None
+
+        if hasattr(self.message, 'delivery_tag') and self.message.delivery_tag:
```

Comment:
I don't think we need this `hasattr` check any more - I believe that was for back-compat with previous uamqp release that don't support the delivery tag

Reply:
updated!